rfc/0006-kubewarden-versioning.md

```diff
@@ -0,0 +1,203 @@
+> work together. They just need to care about running a specific
+> Kubewarden version.
+
+> Users are able to easily understand that a new version of Kubewarden
+> is available, and that more than one component might be upgraded in
+> order to move to the new Kubewarden version.
+
+> A user has an issue with Kubewarden. Given its multiple component
+> nature, we would like to know what was the Kubewarden version that
+> got installed, and through what versions it was upgraded.
+
+# Detailed design
+[design]: #detailed-design
+
+## Many components, one experience
+
+Kubewarden core components, which are exposed to the user are:
+
+- Helm Charts
+- Kubewarden Controller
+- Policy Server
+- kwctl
+
+As a second dimension, there are also policies and their respective
+SDK's. This proposal is going to ignore this dimension in purpose,
+because Kubewarden runtime components (`policy-server` and `kwctl`)
+should be compatible with policies previously built, even with older
+language SDK's.
+
+### Public interfaces
+
+Our public interfaces with users are as follows:
+
+- Helm Charts
+  - `values.yaml`
+- Kubewarden Controller
+  - CLI args
+  - Custom Resource Definitions
+    - Versioned
+- Policy Server
+  - CLI args
+  - Configuration
+    - sigstore related files
+    - Usually handled by the Kubewarden Controller. Regardless we
+      should treat this as public interface, given some users might
+      run the Policy Server independent of the Kubewarden Controller,
+      and might configure it with the configuration file.
+- kwctl
+  - CLI args
+  - Configuration
+    - `sources.yaml`
+    - `docker.json`
+    - sigstore related files
+
+There is also the `waPC` contract between the `policy-evaluator`
+(consumed by the Policy Server and kwctl) and Policies and SDK's.
+
+## Kubewarden toplevel version
+
+This proposal envisions using [`CalVer`](https://calver.org/) to
+version the Kubewarden project.
+
+Had Kubewarden be released today, it would be `Kubewarden
+22.03`. This makes it trivial to understand how old your version of
+Kubewarden is, and how much your installation is drifting from the
+latest releases.
+
+## Kubewarden component versions
+
+Every component in Kubewarden might have its own versioning. The
+Kubewarden maintainers are going to maintain a list of all releases.
+
+This list of releases looks as follows:
+
+```yaml
+"22.03":
+  - component: kubewarden-crds-chart
+    version: 0.1.1
+  - component: kubewarden-controller-chart
+    version: 0.3.6
+  - component: kubewarden-controller
+    version: v0.4.5
+  - component: policy-server
+    version: v0.2.6
+  - component: kwctl
+    version: v0.2.5
+"22.05":
+  - component: kubewarden-crds-chart
+    version: 0.2.0
+  - component: kubewarden-controller-chart
+    version: 0.4.0
+  - component: kubewarden-controller
+    version: v0.4.8
+  - component: policy-server
+    version: v0.3.0
+  - component: kwctl
+    version: v0.3.1
+...
+```
```

Comment:
If a user wants to install `v22.05` in this example, he need to look into this list and then install `kubewarden-crds-chart` `v0.2.0` and `kubewarden-controller-chart` `v0.4.0`, which then it would install the right version of the controller and policy-server, right? I'm thinking it might be easier for user if we also use `v22.05` for the helm charts. Then it would be clear which version they are installing, and the version could be check with a `helm ls`

Reply:
We might have a "toplevel" Kubewarden chart. This chart might be autogenerated based on this umbrella version.

So that we have the global Kubewarden chart, that depends on the specific semver-versioned charts kubewarden-crds and kubewarden-controller charts. However, those charts are still semver versioned. Only the "toplevel" Kubewarden chart would be calver versioned.

Reply:
I like the idea of the toplevel chart